core/src/main/java/org/apache/iceberg/rest/HTTPClient.java

```diff
@@ -0,0 +1,283 @@
+        throwFailure(response, responseBody, errorHandler);
+      }
+
+      if (responseBody == null) {
+        throw new RESTException(
+            "Invalid (null) response body for request (expected %s): method=%s, path=%s, status=%d",
+            responseType.getSimpleName(), method.name(), path, response.getCode());
+      }
+
+      try {
+        return mapper.readValue(responseBody, responseType);
+      } catch (JsonProcessingException e) {
+        throw new RESTException(
+            "Received a success response code of %d, but failed to parse response body into %s",
+            response.getCode(), responseType.getSimpleName());
+      }
+    } catch (IOException e) {
+      throw new RESTException(e, "Error occurred while processing %s request", method);
+    }
+  }
+
+  @Override
+  public void head(String path, Consumer<ErrorResponse> errorHandler) {
+    execute(Method.HEAD, path, null, null, errorHandler);
+  }
+
+  @Override
+  public <T> T get(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    return execute(Method.GET, path, null, responseType, errorHandler);
+  }
+
+  @Override
+  public <T> T post(String path, Object body, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    return execute(Method.POST, path, body, responseType, errorHandler);
+  }
+
+  @Override
+  public <T> T delete(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    return execute(Method.DELETE, path, null, responseType, errorHandler);
+  }
+
+  private void addRequestHeaders(HttpUriRequest request) {
+    request.setHeader(HttpHeaders.ACCEPT, ContentType.APPLICATION_JSON.getMimeType());
+    // Many systems require that content type is set regardless and will fail, even on an empty bodied request.
+    request.setHeader(HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType());
+    additionalHeaders.forEach(request::setHeader);
+  }
+
+  @Override
+  public void close() throws IOException {
+    httpClient.close(CloseMode.GRACEFUL);
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private final Map<String, String> additionalHeaders = Maps.newHashMap();
+    private String uri;
+    private CloseableHttpClient httpClient;
+    private ObjectMapper mapper;
+
+    private Builder() {
+    }
+
+    private static String asBearer(String token) {
+      return String.format("Bearer %s", token);
+    }
+
+    public Builder uri(String baseUri) {
+      Preconditions.checkNotNull(baseUri, "Invalid uri for http client: null");
+      this.uri = RESTUtil.stripTrailingSlash(baseUri);
+      return this;
+    }
+
+    // Visible for testing
+    public Builder httpClient(CloseableHttpClient client) {
+      this.httpClient = client;
+      return this;
+    }
+
+    public Builder mapper(ObjectMapper objectMapper) {
```

Comment:
I think this should also be visible for testing only. The class should use one configured with all of the rest serializers by default and not expose serialization to the caller.